### PR TITLE
merge encoder and prettyencoder

### PR DESCRIPTION
**Pull request by oli-obk**
github seems to do a better job at difffing than my local diff.
i can squash the commits if preferred

### CODE REVIEW

**Comment by alexcrichton** [comment]
Have you measured the perf impact of unifying the two encoders into one?

**Comment by oli-obk** [comment]
separated
```
test base64::tests::bench_from_base64   ... bench:       765 ns/iter (+/- 84) = 266 MB/s
test base64::tests::bench_to_base64     ... bench:       356 ns/iter (+/- 46) = 424 MB/s
test hex::tests::bench_from_hex         ... bench:      1399 ns/iter (+/- 183) = 215 MB/s
test hex::tests::bench_to_hex           ... bench:       356 ns/iter (+/- 11) = 424 MB/s
test json::tests::bench_large           ... bench:      1418 ns/iter (+/- 42)
test json::tests::bench_small           ... bench:      1904 ns/iter (+/- 185)
test json::tests::bench_streaming_large ... bench:       621 ns/iter (+/- 39)
test json::tests::bench_streaming_small ... bench:      1009 ns/iter (+/- 97)
```

merged
```
test base64::tests::bench_from_base64   ... bench:       793 ns/iter (+/- 13) = 257 MB/s
test base64::tests::bench_to_base64     ... bench:       362 ns/iter (+/- 5) = 417 MB/s
test hex::tests::bench_from_hex         ... bench:      1431 ns/iter (+/- 33) = 211 MB/s
test hex::tests::bench_to_hex           ... bench:       365 ns/iter (+/- 11) = 413 MB/s
test json::tests::bench_large           ... bench:      1443 ns/iter (+/- 61)
test json::tests::bench_small           ... bench:      1932 ns/iter (+/- 97)
test json::tests::bench_streaming_large ... bench:       624 ns/iter (+/- 98)
test json::tests::bench_streaming_small ... bench:      1026 ns/iter (+/- 66)

```

slightly slower. also, the benchmarks don't blackbox the choice between pretty or compact. Therefore the optimizer might optimize out a lot, not sure how good it is at that considering all the closures passed around.

File: src/json.rs [hunk 1]
```diff
@@ -491,255 +477,61 @@ macro_rules! emit_enquoted_if_mapkey {
...
+                curr_indent: 0,
+                indent: 2,
+            },
+            is_emitting_map_key: false,
+        }
+    }
+
+    /// Creates a new encoder whose output will be written in compact
+    /// JSON to the specified writer
+    pub fn new_compact(writer: &'a mut fmt::Writer) -> Encoder<'a> {
```
Comment (alexcrichton) [inline src/json.rs]: I think it should be fine to call this one `fn new` instead of `fn new_compact`

**Comment by alexcrichton** [comment]
Ok, I'm happy with those numbers, thanks @oli-obk!

**Comment by oli-obk** [comment]
rebased and new_compact -> new rename

**Comment by alexcrichton** [comment]
Thanks!

**Pull request merged**